src/main/java/net/fabricmc/fabric/api/config/FabricConfigManager.java

```diff
@@ -0,0 +1,137 @@
+package net.fabricmc.fabric.api.config;
+
+import blue.endless.jankson.Jankson;
+import blue.endless.jankson.JsonElement;
+import blue.endless.jankson.JsonObject;
+import blue.endless.jankson.impl.SyntaxError;
+import net.fabricmc.loader.api.FabricLoader;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class FabricConfigManager {
```

Comment:
Ok, firstly thanks for getting things moving on the config front again, this is a lot better than arguing.

jankson is a good choice imo, I know some people wont like it, so please speak up if you have something against this.

However this API is missing quite a few things that I think would are critical.

- There is no way for other mods to get the config from other mods. For example if ModMenu wanted to add a gui config this is currently not possible. I would solve this by having mod configs, and a way to get all configs. Imo mods should be able to have more than one config if they want, in that case the configs should be placed in a directory.

- This doesnt really follow the api/impl structure that we have been moving to, should be an easy fix.

Overall I think its a good start, but having some form of central list of mod's configs would be a requirement. I dont think we need to implement client -> server syncing at this stage.

Reply:
Trying to think of how to make a good map of configs. The Supplier stuff should help when Falk's added the stuff to Jankson (at least I think that's what he meant in his comment?), but would just a map of `String modId, Pair<Class configClass, Object config>` work?

Reply:
Here is what I came up with very quickly. Hopefully this would make more sense than me trying to explain it. (Im not great at that)

```java
import java.io.File;
import java.lang.reflect.InvocationTargetException;
import java.util.HashMap;
import java.util.Map;

class Scratch {

	public static void main(String[] args) {
		//Super basic config
		{
			// ./config/modid.json
			Object simpleConfig = ConfigBuilder.builder("modid").getConfig(Object.class);
		}

		//A mod with more than one config, in sub dirs
		{
			ConfigBuilder configBuilder = ConfigBuilder.builder("modid2");
			// ./config/modid2/client.json
			Object clientConfig = configBuilder.getConfig(Object.class, "client");
			// ./config/modid2/common.json
			Object commonConfig = configBuilder.getConfig(Object.class, "common");
		}

		//Example of trying to create more than one builder per mod, should crash
		{
			ConfigBuilder.builder("modid3").getConfig(Object.class);
			//Crash as above we allready created the basic config
			ConfigBuilder.builder("modid3").getConfig(Object.class, "test");
		}

		//Example of creating a basic config, and then trying to create a named config, should crash
		{
			ConfigBuilder anotherBuilder = ConfigBuilder.builder("modid4");
			anotherBuilder.getConfig(Object.class);
			anotherBuilder.getConfig(Object.class, "crash");
		}


	}

	public static <T> T getConfig(String modID){
		return null;
	}

	/**
	 * This would be interal and have some methods exposed via an api to get existing configs for other mods.
	 */
	public static class ConfigManager {
		//Make this internal
		public static Map<String, ConfigBuilder> modConfigs = new HashMap<>();

		//Add some apis to get configs from a mod id, this would allow other mods to access the config.
	}


	public static class ConfigBuilder {

		private String modId;

		private static Map<String, Object> configs = new HashMap<>();

		//True when only one config file is being used, then defaults to /config/modid.json5
		private boolean singleConfig = false;

		private ConfigBuilder(String modId) {
			if(ConfigManager.modConfigs.containsKey(modId)){
				throw new UnsupportedOperationException("Cannot have more than one config manager per mod : " + modId);
			}
			ConfigManager.modConfigs.put(modId, this);
			this.modId = modId;
		}

		public static ConfigBuilder builder(String modId){
			return new ConfigBuilder(modId);
		}

		/**
		 *
		 * Used to get a basic config file for a mod
		 *
		 * @return the instance of the config
		 */
		public <T> T getConfig(Class<T> clazz){
			T config = getConfig(clazz, "config", false);
			singleConfig = true;
			return config;
		}

		public <T> T getConfig(Class<T> clazz, String name){
			return getConfig(clazz, name, true);
		}

		private  <T> T getConfig(Class<T> clazz, String name, boolean subDir){
			if(singleConfig){
				//Crash if a mod tries to create a sub dir config after it already has done so
				throw new UnsupportedOperationException("Config builder created a single config, cannot create more. Specify config name");
			}
			try {
				//TODO do the config jazz here
				File configFile = new File("config/" + (subDir ? "" : modId) + name);
				T config = clazz.getConstructor().newInstance();
				configs.put(name, config);
				return config;
			} catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
				throw new RuntimeException("Failed to load config class");
			}
		}
	}

}
```